### PR TITLE
terraform workshop blog post

### PR DESCRIPTION
**Pull request by edunham**
todo before merging this: 
- [ ] get Jeff review

todo next week after merging this: 
- [ ] add video link

### CODE REVIEW

File: _source/_posts/2023-07-20-terraform-workshop.md [hunk 1]
```diff
@@ -0,0 +1,558 @@
...
+by: advocate
+communities: [devops,go]
+description: "Get started managing your Okta resources with Terraform"
+tags: [terraform,devops,go,automation]
+tweets:
+- ""
+- ""
+- ""
+image:
+type: awareness|conversion
```
Comment (edunham) [inline _source/_posts/2023-07-20-terraform-workshop.md]:
```suggestion
type: awareness|conversion
```
```suggestion
type: awareness
```

File: _source/_posts/2023-07-20-terraform-workshop.md [hunk 2]
```diff
@@ -0,0 +1,558 @@
+---
+layout: blog_post
+title: "Okta Terraform Workshop"
+author: edunham
+by: advocate
+communities: [devops,go]
+description: "Get started managing your Okta resources with Terraform"
+tags: [terraform,devops,go,automation]
```
Comment (edunham) [inline _source/_posts/2023-07-20-terraform-workshop.md]:
```suggestion
tags: [devops,go,automation]
```

**Comment by edunham** [comment]
I don't know quite what was wrong with the front matter before, but it's working now.

File: _source/_posts/2023-07-28-terraform-workshop.md [hunk 3]
```diff
@@ -0,0 +1,576 @@
...
+# Terraform Formatting
+
+If you're using an editor or IDE that doesn't automatically fix your indentation and other formatting details, run the command `terraform fmt` from the command line in your Terraform project directory to clean up the formatting of all `.tf` files.
+
+Try adding some ugly whitespace to one of your Terraform files, and then use `terraform fmt` to clean it up!
+
+# Importing Resources
+
+What if a user existed in Okta before you started working with Terraform, and you wanted to create a Terraform resource to represent them? [Terraform import](https://developer.hashicorp.com/terraform/language/import), an experimental feature, does this. 
+
```
Comment (edunham) [inline _source/_posts/2023-07-28-terraform-workshop.md:422]:
```suggestion

## Manually create user
```

File: _source/_posts/2023-07-28-terraform-workshop.md [hunk 4]
```diff
@@ -0,0 +1,576 @@
...
+# Key Terraform Concepts
+
+Terraform code describes a desired state for cloud infrastructure. After you write your Terraform code, the `terraform plan` command predicts what changes will be required to make the infrastructure match the code, and the `terraform apply` command changes the infrastructure. 
+
+A Terraform project is a directory containing .tf files. These Terraform files contain providers, resources, and perhaps variables and data sources too. After running `terraform init` in a Terraform project, Terraform downloads all required providers to the `.terraform` directory in that project. 
+
+Terraform also stores data about the state of the infrastructure by creating the file `terraform.tfstate` in the project. Every piece of infrastructure should be represented in only one `terraform.tfstate` file, so if several collaborators work on the same Terraform codebase together, it's best to use Terraform Cloud or a similar service to share a single `terraform.tfstate` file. 
+
+
+# Setting Up the Okta Terraform Provider
```
Comment (edunham) [inline _source/_posts/2023-07-28-terraform-workshop.md]:
```suggestion
# Set Up the Okta Terraform Provider
```

File: _source/_posts/2023-07-28-terraform-workshop.md [hunk 5]
```diff
@@ -0,0 +1,597 @@
...
+# Key Terraform Concepts
+
+Terraform code describes a desired state for cloud infrastructure. After you write your Terraform code, the `terraform plan` command predicts what changes will be required to make the infrastructure match the code, and the `terraform apply` command changes the infrastructure. 
+
+A Terraform project is a directory containing .tf files. These Terraform files contain providers, resources, and perhaps variables and data sources too. After running `terraform init` in a Terraform project, Terraform downloads all required providers to the `.terraform` directory in that project. 
+
+Terraform also stores data about the state of the infrastructure by creating the file `terraform.tfstate` in the project. Every piece of infrastructure should be represented in only one `terraform.tfstate` file, so if several collaborators work on the same Terraform codebase together, it's best to use Terraform Cloud or a similar service to share a single `terraform.tfstate` file. 
+
+
+# Set Up the Okta Terraform Provider
```
Comment (edunham) [inline _source/_posts/2023-07-28-terraform-workshop.md:40]:
```suggestion
## Set Up the Okta Terraform Provider
```

File: _source/_posts/2023-07-28-terraform-workshop.md [hunk 6]
```diff
@@ -0,0 +1,597 @@
...
+  scopes      = ["okta.groups.manage", "okta.users.manage", "okta.policies.manage"]
+  private_key = ""
+}
+
+```
+All of those fields are mandatory, and you will fill out the missing values in the next steps. The `org_name` will come from the URL of your Okta Developer Account, and the `client_id` and `private_key` will come from the App Integration. 
+
+Scopes describe categories of changes that Terraform is allowed to make with the Okta Provider. This list of scopes will appear twice: Once here in the provider definition, and again in the Okta App Integration. The 3 scopes provided will be enough to do the exercises in this workshop. Later on, you can explore what would happen if the scopes had been set up wrong at this step. 
+
+# Create the Okta App Integration
```
Comment (edunham) [inline _source/_posts/2023-07-28-terraform-workshop.md]:
```suggestion
## Create the Okta App Integration
```

File: _source/_posts/2023-07-28-terraform-workshop.md [hunk 7]
```diff
@@ -0,0 +1,597 @@
...
+
+Save your changes to the general settings of the API Service App, and navigate to the app's Okta API Scopes. Grant `okta.groups.manage`, `okta.policies.manage`, and `okta.users.manage` for this workshop. 
+
+In the terminal, `cd okta-terraform-workshop` and convert the PKCS-1 key to an RSA key: 
+
+```
+$ openssl rsa -in pkcs.pem -out rsa.pem
+```
+
+# Configure the Okta Terraform Provider
```
Comment (edunham) [inline _source/_posts/2023-07-28-terraform-workshop.md]:
```suggestion
## Configure the Okta Terraform Provider
```

File: _source/_posts/2023-07-28-terraform-workshop.md [hunk 8]
```diff
@@ -0,0 +1,653 @@
+---
+layout: blog_post
+title: "Okta Terraform Workshop"
```
Comment (edunham) [inline _source/_posts/2023-07-28-terraform-workshop.md]:
```suggestion
title: "Enterprise-Ready Workshop: Terraform"
```

**Comment by edunham** [comment]
for the tf expert feedback requirement, I ran it by Mike Mondragon, and integrated his Slack feedback into the post to address some common questions that beginners to Okta's Terraform ask

File: _source/_posts/2023-07-28-terraform-workshop.md [hunk 9]
```diff
@@ -0,0 +1,660 @@
...
+  scopes      = ["okta.groups.manage", "okta.users.manage", "okta.policies.manage"]
+  private_key = "${path.module}/rsa.pem"
+}
+```
+
+# Manage Users and Groups
+
+With the provider configured, Terraform is ready to make changes to your Okta org. In the Okta Admin Console in your browser, take one last look at the empty Users list (found under the Directory heading in the left sidebar), because you're about to create some users with Terraform!
+
+In keeping with the Terraform theme, this workshop's examples will create users for some plants and animals, and assign them to a garden group. The exercises will work just as well if you choose a different theme, however, so feel free to customize your users and groups to keep things interesting.
```
Comment (ghost) [inline _source/_posts/2023-07-28-terraform-workshop.md]:
```suggestion
In keeping with the Terraform theme, this workshop's examples will create users for some plants and animals, and assign them to a garden group. However, the exercises will work just as well if you choose a different theme, so feel free to customize your users and groups to keep things interesting. 
```

File: _source/_posts/2023-07-28-terraform-workshop.md [hunk 10]
```diff
@@ -0,0 +1,660 @@
...
+  description = <<POEM
+When daisies pied and violets blue
+And lady-smocks all silver-white
+And cuckoo-buds of yellow hue
+Do paint the meadows with delight.  
+POEM  
+}
+```
+
+Notice how `<<` and a custom delimiter, in this case `POEM`, start the multi-line string. The same delimiter, in this case `POEM`, on its own line, ends the string.
```
Comment (ghost) [inline _source/_posts/2023-07-28-terraform-workshop.md]:
```suggestion
Notice how `<<` and a custom delimiter, in this case `POEM`, start the multi-line string. The same delimiter, in this case, `POEM` on its own line ends the string. 
```

**Review by ghost — CHANGES_REQUESTED** [review]

**Review by palermo4 — APPROVED** [review]

File: _source/_posts/2023-07-28-terraform-workshop.md [hunk 11]
```diff
@@ -0,0 +1,660 @@
...
+
+Never run `terraform destroy` on infrastructure that you want to keep! 
+
+# Next Steps
+
+In this workshop, you've taken your first steps managing Okta with Terraform. You've learned to create, change, import, and destroy resources, as well as refactoring and formatting your Terraform code. You've also begun to consider more advanced topics that you'll encounter in your Terraform journey, like config drift and API usage. 
+
+* Learn more about Terraform in the Hashicorp docs. Could Terraform help you manage cloud infrastructure beyond Okta?
+* Explore other resource types in the Okta Terraform Provider docs, and try configuring them in your developer account. 
+* If your organization already uses Terraform to manageinfrastructure, learn about how terraform.tfstate files and secrets are managed.
```
Comment (isemona) [inline _source/_posts/2023-07-28-terraform-workshop.md:658]:
```
suggestion
manageinfrastructure to manage infrastructure
```

**Pull request merged**